server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -91,63 +118,257 @@ private void registerListeners() {
+        DiscoveryNode previousMasterNode = event.previousState().nodes().getMasterNode();
+        if (currentMasterNode == null || previousMasterNode == null) {
+            return currentMasterNode != previousMasterNode;
         }
+        return previousMasterNode.getEphemeralId().equals(currentMasterNode.getEphemeralId()) == false;
+    }
+
+    // We compare the current health node against both the last seen health node from this node and the
+    // health node reported in the previous cluster state to be safe that we do not miss any change due to
+    // a flaky state.
+    private boolean hasHealthNodeChanged(DiscoveryNode currentHealthNode, ClusterChangedEvent event) {
+        DiscoveryNode previousHealthNode = HealthNode.findHealthNode(event.previousState());
+        return Objects.equals(lastSeenHealthNode.get(), currentHealthNode == null ? null : currentHealthNode.getId()) == false
+            || Objects.equals(previousHealthNode, currentHealthNode) == false;
     }
 
+    @Nullable
     DiskHealthInfo getLastReportedDiskHealthInfo() {
-        return lastReportedDiskHealthInfo;
+        return lastReportedDiskHealthInfo.get();
+    }
+
+    /**
+     * This class is responsible for running the health monitoring. It evaluates and checks the health info of this node
+     * in the configured intervals. The first run happens upon initialization. If there is an exception, it will log it
+     * and continue to schedule the next run.
+     */
+    static class Monitoring implements Runnable, Scheduler.Cancellable {
+
+        private final TimeValue interval;
+        private final String executor;
+        private final Scheduler scheduler;
+        private final ClusterService clusterService;
+        private final DiskCheck diskCheck;
+        private final Client client;
+
+        private final AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo;
+        private final AtomicReference<String> lastSeenHealthNode;
+
+        private volatile boolean cancelled = false;
+        private volatile Scheduler.ScheduledCancellable scheduledRun;
+
+        Monitoring(
+            TimeValue interval,
+            Scheduler scheduler,
+            AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo,
+            AtomicReference<String> lastSeenHealthNode,
+            DiskCheck diskCheck,
+            ClusterService clusterService,
+            Client client
+        ) {
+            this.interval = interval;
+            this.executor = ThreadPool.Names.MANAGEMENT;
+            this.scheduler = scheduler;
+            this.lastReportedDiskHealthInfo = lastReportedDiskHealthInfo;
+            this.lastSeenHealthNode = lastSeenHealthNode;
+            this.clusterService = clusterService;
+            this.diskCheck = diskCheck;
+            this.client = client;
+            scheduledRun = scheduler.schedule(this, TimeValue.ZERO, executor);
```

Comment:
`this` is incomplete as it's in the constructor.
Shall we have a start/schedule/init method that kicks things off?

Reply:
Yes, same as we do with the cluster state listeners. On it.